attestation/slsa/slsa.go

```diff
@@ -0,0 +1,199 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package slsa
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	prov "github.com/in-toto/attestation/go/predicates/provenance/v1"
+	v1 "github.com/in-toto/attestation/go/v1"
+	"github.com/in-toto/go-witness/attestation"
+	"github.com/in-toto/go-witness/attestation/commandrun"
+	"github.com/in-toto/go-witness/attestation/environment"
+	"github.com/in-toto/go-witness/attestation/git"
+	"github.com/in-toto/go-witness/attestation/material"
+	"github.com/in-toto/go-witness/attestation/product"
+	"github.com/in-toto/go-witness/cryptoutil"
+	"github.com/in-toto/go-witness/registry"
+	"google.golang.org/protobuf/types/known/structpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	Name    = "slsa"
+	Type    = "https://slsa.dev/provenance/v1.0"
+	RunType = attestation.PostProductRunType
+
+	defaultExport = false
+)
+
+// This is a hacky way to create a compile time error in case the attestor
+// doesn't implement the expected interfaces.
+var (
+	_ attestation.Attestor  = &Provenance{}
+	_ attestation.Subjecter = &Provenance{}
+)
+
+func init() {
+	attestation.RegisterAttestation(Name, Type, RunType,
+		func() attestation.Attestor { return New() },
+		registry.BoolConfigOption(
+			"export",
+			"Export the SLSA provenance attestation to its own file",
+			defaultExport,
+			func(a attestation.Attestor, export bool) (attestation.Attestor, error) {
+				slsaAttestor, ok := a.(*Provenance)
+				if !ok {
+					return a, fmt.Errorf("unexpected attestor type: %T is not a SLSA provenance attestor", a)
+				}
+				WithExport(export)(slsaAttestor)
+				return slsaAttestor, nil
+			},
+		),
+	)
+}
+
+type Option func(*Provenance)
+
+func WithExport(export bool) Option {
+	return func(p *Provenance) {
+		p.export = export
+	}
+}
+
+type Provenance struct {
+	PbProvenance prov.Provenance
+	products     map[string]attestation.Product
+	export       bool
+}
+
+func New() *Provenance {
+	return &Provenance{}
+}
+
+func (p *Provenance) Name() string {
+	return Name
+}
+
+func (p *Provenance) Type() string {
+	return Type
+}
+
+func (p *Provenance) RunType() attestation.RunType {
+	return RunType
+}
+
+func (p *Provenance) Export() bool {
+	return p.export
+}
+
+func (p *Provenance) Attest(ctx *attestation.AttestationContext) error {
+	builder := prov.Builder{}
+	metadata := prov.BuildMetadata{}
+	p.PbProvenance.BuildDefinition = &prov.BuildDefinition{}
+	p.PbProvenance.RunDetails = &prov.RunDetails{Builder: &builder, Metadata: &metadata}
+
+	p.PbProvenance.BuildDefinition.BuildType = "https://witness.dev/slsa-build@v0.1"
+	p.PbProvenance.RunDetails.Builder.Id = "https://witness.dev/witness-github-action@v0.1"
```

Comment:
@jkjell I think i'm right in saying that we need to add extra custom logic so that we can add this metadata depending on where / how witness was invocated? Maybe something that we might need to think about.

The way I see it, we should (as a first pass) look for attestors such as github, gitlab, google cloud build? If those attestors have run and are populated, we can use those to populate these fields.

Reply:
Yep, GitHub and GitLab are the two we'll implement for sure. We'll have to come up with a reasonable name for a default or "undetected".